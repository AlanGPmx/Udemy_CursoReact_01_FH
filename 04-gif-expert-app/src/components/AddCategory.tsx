import {useState} from 'react';
import {AddCategoryInterface} from '../interfaces/AddCategory.interface';
import PropTypes from 'prop-types';

export const AddCategory = (props: AddCategoryInterface) => {
	const {onNewCategory} = props;
	const [inputValue, setInputValue] = useState('');

	const onInputChange = (event: any) => {
		const {target} = event;
		const {value} = target;
		setInputValue(value);
	};

	const onSubmit = (event: any) => {
		event.preventDefault();
		const newValue = inputValue.trim();

		if (newValue.length <= 1) return;

		setInputValue('');
		onNewCategory(newValue);
	};

	return (
		<form onSubmit={onSubmit} aria-label='form'>
			<input
				type='text'
				placeholder='Busca un GIF...'
				value={inputValue}
				onChange={onInputChange}
			/>
		</form>
	);
};

AddCategory.propTypes = {
	onNewCategory: PropTypes.func.isRequired,
};
