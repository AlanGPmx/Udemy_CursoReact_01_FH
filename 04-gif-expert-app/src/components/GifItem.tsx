import {GifItemInterface} from '../interfaces/GiftItem.interface';
import PropTypes from 'prop-types';

export const GifItem = (props: GifItemInterface) => {
	const {title, url} = props;
	return (
		<div className='card'>
			<img src={url} alt={title} />
			<p>{title}</p>
		</div>
	);
};

// Es la alternativa en JS y React al tipado que TS ofrece
GifItem.propTypes = {
	title: PropTypes.string.isRequired,
	url: PropTypes.string.isRequired,
};

//La alternativa para asignar un valor por defecto a las Props y no hacerlo en crudo en los argumentos del componente
GifItem.defaultProps = {
	url: 'https://www.publicdomainpictures.net/pictures/280000/nahled/not-found-image-15383864787lu.jpg',
	title: 'Imagen no encontrada',
};
