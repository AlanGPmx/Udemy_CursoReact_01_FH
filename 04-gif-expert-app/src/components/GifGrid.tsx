import {useFetchGifs} from '../hooks/useFetchGifs';
import {GifGridInterface} from '../interfaces/GifGridInterface';
import {GifItem} from './GifItem';
import PropTypes from 'prop-types';

export const GifGrid = (props: GifGridInterface) => {
	const {category} = props;

	const {gifs, isLoading} = useFetchGifs(category);

	return (
		<>
			<h3>{category}</h3>
			{isLoading && <h2>Cargando...</h2>}
			<div className='card-grid'>
				{gifs.map((_gif: any) => (
					<GifItem key={_gif.id} {..._gif} />
				))}
			</div>
		</>
	);
};

GifGrid.propTypes = {
	category: PropTypes.string.isRequired,
};
