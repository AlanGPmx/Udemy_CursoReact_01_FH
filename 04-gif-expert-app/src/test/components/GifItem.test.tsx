import {render, screen} from '@testing-library/react';
import {GifItem} from '../../components/GifItem';

describe('Evaluando el componente GifItem', () => {
	const title = 'Club América';
	const url = 'https://media.giphy.com/media/3o7TKsQ8U5h3iXWgZK/giphy.gif';
	const id = '14';

	test('Deberá renderizar el componente y compararlo con el Snapshot', () => {
		const {container} = render(<GifItem title={title} url={url} id={id} />);
		expect(container).toMatchSnapshot();
	});

	test('Deberá mostrar el título y la imagen', () => {
		render(<GifItem title={title} url={url} id={id} />);
		//screen.debug();

		// Opcion 1: Menos limpia
		// expect(screen.getByRole('img').src).toBe(url);
		// expect(screen.getByRole('img').alt).toBe(title);

		//Option 2: Más limpia
		const {src, alt} = screen.getByRole('img') as HTMLInputElement;
		expect(src).toBe(url);
		expect(alt).toBe(title);
	});

	test('Deberá mostrar el título del componente', () => {
		render(<GifItem title={title} url={url} id={id} />);
		expect(screen.getByText(title)).toBeTruthy();
	});
});
