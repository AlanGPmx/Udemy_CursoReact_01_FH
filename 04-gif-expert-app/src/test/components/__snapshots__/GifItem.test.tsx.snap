// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Evaluando el componente GifItem Deberá renderizar el componente y compararlo con el Snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Club América"
      src="https://media.giphy.com/media/3o7TKsQ8U5h3iXWgZK/giphy.gif"
    />
    <p>
      Club América
    </p>
  </div>
</div>
`;
